server/service/sessions.go

```diff
@@ -594,3 +595,82 @@ func (svc *Service) validateSession(ctx context.Context, session *fleet.Session)
 
 	return svc.ds.MarkSessionAccessed(ctx, session)
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Demo Login
+////////////////////////////////////////////////////////////////////////////////
+
+// This is a special kind of login where the username and password come in form values rather than
+// JSON as is typical for API requests. This is intended to support logins from demo environments,
+// when users are being redirected from fleetdm.com.
+
+type demologinRequest struct {
+	Email    string
+	Password string
+}
+
+func (demologinRequest) DecodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
+	err := r.ParseForm()
+	if err != nil {
+		return nil, ctxerr.Wrap(ctx, &badRequestError{message: err.Error()}, "decode demo login")
+	}
+
+	return demologinRequest{
+		Email:    r.FormValue("email"),
+		Password: r.FormValue("password"),
+	}, nil
+}
+
+type demologinResponse struct {
+	content string
+	Err     error `json:"error,omitempty"`
+}
+
+func (r demologinResponse) error() error { return r.Err }
+
+// If html is present we return a web page
+func (r demologinResponse) html() string { return r.content }
+
+func makeDemologinEndpoint(urlPrefix string) handlerFunc {
+	return func(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+		if os.Getenv("FLEET_DEMO") != "1" {
+			return nil, errors.New("this endpoint only enabled in demo mode")
+		}
+		req := request.(demologinRequest)
+
+		_, sess, err := svc.Login(ctx, req.Email, req.Password)
+
+		session := struct {
+			Token string
+		}{}
+		var resp demologinResponse
+		if err != nil {
+			resp.Err = err
+		}
+		if sess != nil {
+			session.Token = sess.Key
+		}
+
+		relayStateLoadPage := `<!DOCTYPE html>
+     <script type='text/javascript'>
+     window.localStorage.setItem('FLEET::auth_token', '{{ .Token }}');
+     window.location = "/";
+     </script>
+     <body>
+     Redirecting to Fleet...
+     </body>
+     </html>
+    `
+		tmpl, err := template.New("demologin").Parse(relayStateLoadPage)
+		if err != nil {
+			return nil, err
```

Comment:
The convention in these `make*Endpoint` methods I believe is to return:
```go
return demologinResponse{Err: err}, nil
```

Reply:
Ah yes but this one is slightly different because we want to return HTML with a redirect. It's most similar to the SSO callback: https://github.com/fleetdm/fleet/blob/main/server/service/sessions.go#L348-L382